core/embed/trezorhal/stm32f4/flash.c

```diff
@@ -104,48 +110,220 @@ const void *flash_get_address(uint8_t sector, uint32_t offset, uint32_t size) {
+  uint32_t size = 0;
+  for (int s = 0; s < subarea->num_sectors; s++) {
+    size += flash_sector_size(subarea->first_sector + s);
+  }
+  return size;
+}
+
+uint32_t flash_area_get_size(const flash_area_t *area) {
+  uint32_t size = 0;
+  for (int i = 0; i < area->num_subareas; i++) {
+    size += flash_subarea_get_size(&area->subarea[i]);
+  }
+  return size;
+}
+
+uint16_t flash_total_sectors(const flash_area_t *area) {
+  uint16_t total = 0;
+  for (int i = 0; i < area->num_subareas; i++) {
+    total += area->subarea[i].num_sectors;
+  }
+  return total;
+}
+
+int flash_get_sector_num(const flash_area_t *area, uint32_t sector_inner_num) {
+  uint16_t sector = 0;
+  uint16_t remaining = sector_inner_num;
+  for (int i = 0; i < area->num_subareas; i++) {
+    if (remaining < area->subarea[i].num_sectors) {
+      sector = area->subarea[i].first_sector + remaining;
+      return sector;
+    } else {
+      remaining -= area->subarea[i].num_sectors;
+    }
+  }
+
+  return -1;
+}
+
+int32_t flash_get_sector(const flash_area_t *area, uint32_t offset) {
+  uint32_t tmp_offset = offset;
+  for (int i = 0; i < area->num_subareas; i++) {
+    uint16_t sector = area->subarea[i].first_sector;
+
+    uint32_t sub_size = flash_subarea_get_size(&area->subarea[i]);
+    if (tmp_offset >= sub_size) {
+      tmp_offset -= sub_size;
+      continue;
+    }
+
+    // in correct subarea
+    for (int s = 0; s < area->subarea[i].num_sectors; s++) {
+      const uint32_t sector_size = flash_sector_size(sector);
+      if (tmp_offset >= sector_size) {
+        tmp_offset -= sector_size;
+        sector++;
+
+        if (s == area->subarea[i].num_sectors - 1) {
+          return -1;
+        }
+        continue;
+      }
+      // in correct sector
+      return sector;
+    }
+  }
+  return -1;
+}
+
+const void *flash_area_get_address(const flash_area_t *area, uint32_t offset,
+                                   uint32_t size) {
+  uint32_t tmp_offset = offset;
+
+  for (int i = 0; i < area->num_subareas; i++) {
+    uint16_t sector = area->subarea[i].first_sector;
+
+    uint32_t sub_size = flash_subarea_get_size(&area->subarea[i]);
+    if (tmp_offset >= sub_size) {
+      tmp_offset -= sub_size;
+      continue;
+    }
+
+    // in correct subarea
+
+    for (int s = 0; s < area->subarea[i].num_sectors; s++) {
+      const uint32_t sector_size = flash_sector_size(sector);
+      if (tmp_offset >= sector_size) {
+        tmp_offset -= sector_size;
+        sector++;
+
+        if (s == area->subarea[i].num_sectors - 1) {
+          return NULL;
+        }
+        continue;
+      }
+      // in correct sector
+      break;
+    }
+
+    const void *addr = flash_get_address(sector, tmp_offset, size);
+    const void *area_end = flash_get_address(sector + 1, 0, 0);
+    if ((uint32_t)addr + size > (uint32_t)area_end) {
+      return NULL;
+    }
+    return (const void *)addr;
+  }
+  return NULL;
+}
+
+secbool flash_area_erase(const flash_area_t *area,
```

Comment:
Can this function just call `flash_area_erase_bulk` with `count = 1`?

Reply:
yeah https://github.com/trezor/trezor-firmware/commit/55785b2157d3e49913a1c4140bca60d582d5f63a